### PR TITLE
add test for #78894 #71450

### LINKED ISSUE #71450: False positive unused_import warning on std::string::String
This (very contrived) example emits a warning:

```rust
mod foo {
    pub struct String;
    
    #[derive(Debug)]
    pub struct Number;
}

fn main() {
    use std::string::String;
    use foo::*;
    
    let n = Number;
    let s = String::new();

    println!("{:?} {}", n, s);
}
```

Playground: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=4333aa0dec7b94c3d2942e7380894003

```
warning: the item `String` is imported redundantly
  --> src/main.rs:9:9
   |
9  |     use std::string::String;
   |         ^^^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(unused_imports)]` on by default
```

This is a false positive because removing `use std::string::String` results in a compilation error:

```
error[E0599]: no function or associated item named `new` found for struct `foo::String` in the current scope
  --> src/main.rs:13:21
   |
2  |     pub struct String;
   |     ------------------ function or associated item `new` not found for this
...
13 |     let s = String::new();
   |                     ^^^ function or associated item not found in `foo::String`
```

### Meta
```
$ rustc --version --verbose
rustc 1.42.0 (b8cedc004 2020-03-09)
binary: rustc
commit-hash: b8cedc00407a4c56a3bda1ed605c6fc166655447
commit-date: 2020-03-09
host: x86_64-pc-windows-msvc
release: 1.42.0
LLVM version: 9.0
```

Same for `1.43.0-beta.6` and `1.44.0-nightly (2020-04-21 45d050cde277b22a7558)`.

### LINKED ISSUE #78894: Incorrect `unused_import` lint firing for a function-local `macro_rules!`
I tried this code:

```rust
fn main ()
{
    m!();
    // where
    macro_rules! m {
        () => ();
    } use m;
}
```

I expected to see this happen: no warning.

Instead, this happened: [I get a warning](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=21509e89ac39afe8028e563cbf8530c6) saying:

```text
warning: the item `m` is imported redundantly
 --> src/main.rs:7:11
  |
5 | /     macro_rules! m {
6 | |         () => ();
7 | |     } use m;
  | |     -     ^
  | |_____|
  |       the item `m` is already defined here
  |
  = note: `#[warn(unused_imports)]` on by default
```

This is incorrect since removing that `use` "statement" leads to the code not compiling.

### Meta

Affects all three versions currently available in the Playground (1.47.0 -> 1.49.0)

### PR DESCRIPTION
**Pull request by surechen**
fixes #78894 
fixes #71450

### CODE REVIEW

File: tests/ui/lint/use-redundant/use-redundant-issue-78894.rs [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+fn main ()
+{
+    bar!();
+
+    macro_rules! bar {
+        () => ();
+    }
+
+    use bar;
```
Comment (petrochenkov) [inline tests/ui/lint/use-redundant/use-redundant-issue-78894.rs:13]: Could you add a test case with a macro + import pair at module level as well?

**Comment by petrochenkov** [comment]
r=me with one more test case added.
@rustbot author

**Comment by surechen** [comment]
> r=me with one more test case added. @rustbot author

Thank you, added tests. I am not sure if I understand correctly. Please help me review it again.

**Comment by petrochenkov** [comment]
Thanks!
@bors r+

**Pull request merged**